### PR TITLE
Check total elements before showing separator line

### PR DESCRIPTION
**Pull request by abeddow91**
## What does this change?
Checks if the key takeaways element is the last element in the article. If it isn't, display the separator line at the end of the key takeaways element.

## Why?
This was requested by design.

## Screenshots

| Before      | After      |
| ----------- | ---------- |
| ![before][] | ![after][] |

[before]: https://github.com/guardian/dotcom-rendering/assets/20416599/a7f5827f-61a4-4ebd-a489-15b5b0658445
[after]: https://github.com/guardian/dotcom-rendering/assets/20416599/3f7c1360-669a-44c0-ae3f-3fe997f3118c

### CODE REVIEW

File: dotcom-rendering/src/components/KeyTakeaways.tsx [hunk 1]
```diff
@@ -21,9 +21,11 @@ interface KeyTakeawaysProps {
 	starRating?: number;
 	keyTakeaways: KeyTakeaway[];
 	RenderArticleElement: ArticleElementRenderer;
+	totalElements?: number;
+	index?: number;
 }
 
-const finalLineStyles = css`
```
Comment (abeddow91) [inline dotcom-rendering/src/components/KeyTakeaways.tsx:26]: I've renamed this as it's actually used to separate the key takeaways from further content, not as the final line of key takeaways

File: dotcom-rendering/src/lib/ArticleRenderer.tsx [hunk 2]
```diff
@@ -66,6 +66,7 @@ export const ArticleRenderer = ({
 	abTests,
 	editionId,
 }: Props) => {
+	const totalElements = elements.length;
 	const renderedElements = elements.map((element, index) => {
```
Comment (mxdvl) [inline dotcom-rendering/src/lib/ArticleRenderer.tsx:70]: You could do this inline with the [third parameter of the `map` callback](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map#array):

```ts
const renderedElements = elements.map((element, index, { length } ) => {
  // do stuff
}
```

File: dotcom-rendering/src/components/KeyTakeaways.tsx [hunk 3]
```diff
@@ -21,9 +21,11 @@ interface KeyTakeawaysProps {
 	starRating?: number;
 	keyTakeaways: KeyTakeaway[];
 	RenderArticleElement: ArticleElementRenderer;
+	totalElements?: number;
+	index?: number;
```
[the remark this answers is not in the document]
Comment (abeddow91) [reply dotcom-rendering/src/components/KeyTakeaways.tsx]: Sounds good, i've refactored to use isLastElement

File: dotcom-rendering/src/lib/renderElement.tsx [hunk 4]
```diff
@@ -440,6 +442,7 @@ export const renderElement = ({
 					switches={switches}
 					editionId={editionId}
 					RenderArticleElement={RenderArticleElement}
+					isLastElement={index >= totalElements - 1}
```
Comment (mxdvl) [inline dotcom-rendering/src/lib/renderElement.tsx:445]: Technically if we’re confident enough about our code…?

```suggestion
					isLastElement={index === totalElements - 1}
```

**Review by mxdvl — APPROVED** [review]

File: dotcom-rendering/src/components/KeyTakeaways.tsx [hunk 5]
```diff
@@ -21,9 +21,11 @@ interface KeyTakeawaysProps {
 	starRating?: number;
 	keyTakeaways: KeyTakeaway[];
 	RenderArticleElement: ArticleElementRenderer;
+	// is this the last element in the article?
```
Comment (mxdvl) [inline dotcom-rendering/src/components/KeyTakeaways.tsx:24]: If you use JSDoc here it will be available in IDEs as information about this prop

**Pull request merged**